File: flut/flutter/painting/box_decoration.py
from typing import Optional, override
from flut._flut.engine import FlutValueObject, _flut_pack_value
from flut._flut.runtime import _flut_unpack_optional_field, _flut_unpack_required_field
from .box_border import BoxShape


class BoxDecoration(FlutValueObject):
    _flut_type = "BoxDecoration"

    def __init__(
        self,
        *,
        color=None,
        image=None,
        border=None,
        borderRadius=None,
        boxShadow: Optional[list] = None,
        gradient=None,
        backgroundBlendMode=None,
        shape: BoxShape = BoxShape.rectangle,
    ):
        super().__init__()
        self.color = color
        self.image = image
        self.border = border
        self.borderRadius = borderRadius
        self.boxShadow = boxShadow
        self.gradient = gradient
        self.backgroundBlendMode = backgroundBlendMode
        self.shape = shape

    @staticmethod
    def _flut_unpack(data: dict) -> "BoxDecoration":
        return BoxDecoration(
            color=_flut_unpack_optional_field(data, "color"),
            image=_flut_unpack_optional_field(data, "image"),
            border=_flut_unpack_optional_field(data, "border"),
            borderRadius=_flut_unpack_optional_field(data, "borderRadius"),
            boxShadow=_flut_unpack_optional_field(data, "boxShadow"),
            gradient=_flut_unpack_optional_field(data, "gradient"),
            backgroundBlendMode=_flut_unpack_optional_field(
                data, "backgroundBlendMode"
            ),
            shape=_flut_unpack_required_field(data, "shape"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.color is not None:
            result["color"] = _flut_pack_value(self.color)
        if self.image is not None:
            result["image"] = _flut_pack_value(self.image)
        if self.border is not None:
            result["border"] = _flut_pack_value(self.border)
        if self.borderRadius is not None:
            result["borderRadius"] = _flut_pack_value(self.borderRadius)
        if self.boxShadow is not None:
            result["boxShadow"] = _flut_pack_value(self.boxShadow)
        if self.gradient is not None:
            result["gradient"] = _flut_pack_value(self.gradient)
        if self.backgroundBlendMode is not None:
            result["backgroundBlendMode"] = _flut_pack_value(self.backgroundBlendMode)
        result["shape"] = _flut_pack_value(self.shape)
        return result

File: flut/.flutter/lib/flutter/painting/box_decoration.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/dart/ui.dart';
import 'package:flut/flutter/painting/border_radius.dart';
import 'package:flut/flutter/painting/box_border.dart';
import 'package:flut/flutter/painting/box_shadow.dart';
import 'package:flut/flutter/painting/gradient.dart';

class FlutBoxDecoration extends FlutValueObject {
  final BoxDecoration boxDecoration;

  const FlutBoxDecoration(this.boxDecoration) : super('BoxDecoration');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (boxDecoration.color != null) {
      result['color'] = FlutColor(boxDecoration.color!).flutEncode();
    }
    if (boxDecoration.border != null && boxDecoration.border is Border) {
      result['border'] = FlutBorder(
        boxDecoration.border! as Border,
      ).flutEncode();
    }
    if (boxDecoration.borderRadius != null &&
        boxDecoration.borderRadius is BorderRadius) {
      result['borderRadius'] = FlutBorderRadius(
        boxDecoration.borderRadius! as BorderRadius,
      ).flutEncode();
    }
    if (boxDecoration.boxShadow != null) {
      result['boxShadow'] = boxDecoration.boxShadow!
          .map((s) => FlutBoxShadow(s).flutEncode())
          .toList();
    }
    if (boxDecoration.gradient != null) {
      if (boxDecoration.gradient is LinearGradient) {
        result['gradient'] = FlutLinearGradient(
          boxDecoration.gradient! as LinearGradient,
        ).flutEncode();
      } else if (boxDecoration.gradient is RadialGradient) {
        result['gradient'] = FlutRadialGradient(
          boxDecoration.gradient! as RadialGradient,
        ).flutEncode();
      } else if (boxDecoration.gradient is SweepGradient) {
        result['gradient'] = FlutSweepGradient(
          boxDecoration.gradient! as SweepGradient,
        ).flutEncode();
      }
    }
    return result;
  }

  static BoxDecoration? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    List<BoxShadow>? boxShadow;
    final boxShadowData = data['boxShadow'] as List<dynamic>?;
    if (boxShadowData != null) {
      boxShadow = boxShadowData
          .map(
            (s) =>
                FlutBoxShadow.flutDecode(runtime, s as Map<String, dynamic>)!,
          )
          .toList();
    }
    Gradient? gradient;
    final gradientData = data['gradient'] as Map<String, dynamic>?;
    if (gradientData != null) {
      gradient = runtime.decodeObject<Gradient>(gradientData);
    }
    return BoxDecoration(
      color: runtime.unpackOptionalField<Color>(data, 'color'),
      border: runtime.unpackOptionalField<Border>(data, 'border'),
      borderRadius: runtime.unpackOptionalField<BorderRadius>(
        data,
        'borderRadius',
      ),
      boxShadow: boxShadow,
      gradient: gradient,
      backgroundBlendMode: runtime.unpackOptionalField<BlendMode>(
        data,
        'backgroundBlendMode',
      ),
      shape: runtime.unpackRequiredField<BoxShape>(data, 'shape'),
    );
  }
}
